src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -1017,6 +1017,102 @@ void InstanceKlass::initialize_super_interfaces(TRAPS) {
   }
 }
 
+class InitErrorElement {
+  Symbol* _exception;
+  Symbol* _message;
+  OopHandle _stack_trace;
+  const char* _thread_name;
+ public:
+  InitErrorElement(JavaThread* current, Handle exception, Handle stack_trace) {
+    _exception = exception->klass()->name();
+    _exception->increment_refcount();
+    _message = java_lang_Throwable::detail_message(exception());
+    if (_message != nullptr) {
+      _message->increment_refcount();
+    }
+    _stack_trace = OopHandle(Universe::vm_global(), stack_trace());
+    ResourceMark rm(current);
+    stringStream st;
+    st.print("%s", current->name());
+    _thread_name = st.as_string(/* c_heap */ true);
+  }
+  void clean() {
+    _exception->decrement_refcount();
+    if (_message != nullptr) {
+      _message->decrement_refcount();
+    }
+    _stack_trace.release(Universe::vm_global());
+    FREE_C_HEAP_ARRAY(char*, _thread_name);
+  }
+  Symbol* exception() const { return _exception; }
+  Symbol* message() const   { return _message; }
+  oop stack_trace() const   { return _stack_trace.resolve(); }
+  const char* thread_name() const { return _thread_name; }
+};
+
+ResizeableResourceHashtable<const InstanceKlass*, InitErrorElement, ResourceObj::C_HEAP, mtClass>
+      _initialization_error_table(20);
```

Comment:
As I understand, the whole point of this table is to save space. Instead, you could repurpose `T[] Class.enumConstants` to be `Object Class.enumConstantsOrInitError`. For a non-enum classes, this field is never used (and thus already "wasted") and for enum classes it will never be successfully initialized if there's a clinit error.

Reply:
It seems like if this field in Class is wasted, it should be used by something more frequent and at the Java level.  I don't want to add some internal implementation dependency to this.  It'll complicate any attempts to change it.
In short, I didn't know about it and in fact, I shouldn't know about it from inside the JVM.

Reply:
> I shouldn't know about it from inside the JVM

Why not? The JVM already knows about a bunch of other fields in `java.lang.Class`.

That said, I agree that it's probably best left for something more frequent.